File: flutter_app/core/test/receiver/frames_test.dart
/// Receiver FrameBuffer tests — behavior locked before the module exists
/// (TDD RED). Synthetic frames are built with the core wire encoder so the
/// buffer is exercised exactly as the PWA port specifies.
library;

import 'dart:typed_data';

import 'package:qr_transfer_core/protocol/constants.dart';
import 'package:qr_transfer_core/protocol/metadata.dart';
import 'package:qr_transfer_core/protocol/wire.dart';
import 'package:qr_transfer_core/receiver/frames.dart';
import 'package:test/test.dart';

const String sessionA = 'aaaaaaaaaaaaaaaa';
const String sessionB = 'bbbbbbbbbbbbbbbb';

/// Deterministic per-esi payload so symbol order/identity is checkable.
Uint8List payloadFor(int esi) => Uint8List.fromList(
  List<int>.generate(64, (i) => (esi * 13 + i * 7) & 0xff),
);

/// Synthetic DATA wire frame for a fake session.
Uint8List dataFrame(String sessionId, int esi, {int k = 100}) => encodeFrame(
  Frame(
    type: typeData,
    sessionId: sessionId,
    esi: esi,
    k: k,
    totalLen: 0,
    flags: 0,
    payload: payloadFor(esi),
  ),
);

/// Synthetic META wire frame for a fake session.
Uint8List metaFrame(String sessionId, {int k = 100}) => buildMetadataFrame(
  TransferMetadata(
    magic: metaMagic,
    protoVer: protoVersion,
    sessionId: sessionId,
    filename: 'fake.bin',
    mime: 'application/octet-stream',
    totalSize: 0,
    compressedSize: 0,
    compressed: false,
    k: k,
    symbolSize: 1028,
    mtu: 1028,
    fileSHA256: 'a' * 64,
    flags: 0,
  ),
);

void main() {
  group('FrameBuffer', () {
    test('empty buffer exposes null session state and zero counts', () {
      final buffer = FrameBuffer();
      expect(buffer.metadata, isNull);
      expect(buffer.sessionId, isNull);
      expect(buffer.k, isNull);
      expect(buffer.uniqueSymbolCount, 0);
      expect(buffer.totalFramesSeen, 0);
      expect(buffer.droppedCount, 0);
      expect(buffer.symbols(), isEmpty);
      expect(buffer.symbolEsiSet(), isEmpty);
    });

    test('k is exposed from DATA headers before any META frame', () {
      final buffer = FrameBuffer();
      final result = buffer.feed(dataFrame(sessionA, 0, k: 100));
      expect(result.status, FeedStatus.ok);
      expect(buffer.k, 100);
      expect(buffer.metadata, isNull);
    });

    test('in-order feed of k symbols latches session and stores all', () {
      final buffer = FrameBuffer();
      final first = buffer.feed(dataFrame(sessionA, 0));
      expect(first.status, FeedStatus.ok);
      expect(first.isNewSession, isTrue);
      expect(first.frame, isNotNull);

      for (var esi = 1; esi < 100; esi++) {
        final result = buffer.feed(dataFrame(sessionA, esi));
        expect(result.status, FeedStatus.ok);
        expect(result.isNewSession, isFalse);
      }

      expect(buffer.sessionId, sessionA);
      expect(buffer.uniqueSymbolCount, 100);
      expect(buffer.droppedCount, 0);
      expect(buffer.totalFramesSeen, 100);
      final symbols = buffer.symbols();
      expect(symbols.length, 100);
      expect(symbols.first, payloadFor(0));
      expect(symbols.last, payloadFor(99));
      expect(buffer.symbolEsiSet(), {for (var i = 0; i < 100; i++) i});
    });

    test('out-of-order feed stores the same k distinct symbols', () {
      final buffer = FrameBuffer();
      // (i * 37) % 100 is a permutation of 0..99 since gcd(37, 100) == 1.
      for (var i = 0; i < 100; i++) {
        expect(
          buffer.feed(dataFrame(sessionA, (i * 37) % 100)).status,
          FeedStatus.ok,
        );
      }
      expect(buffer.uniqueSymbolCount, 100);
      expect(buffer.droppedCount, 0);
      expect(buffer.totalFramesSeen, 100);
      final symbols = buffer.symbols();
      expect(symbols.length, 100);
      expect(symbols.first, payloadFor(0));
      expect(symbols.last, payloadFor(99));
    });

    test('duplicate esi feeds are counted but not re-stored', () {
      final buffer = FrameBuffer();
      for (var esi = 0; esi < 100; esi++) {
        buffer.feed(dataFrame(sessionA, esi));
      }
      expect(buffer.feed(dataFrame(sessionA, 5)).status, FeedStatus.ok);
      buffer.feed(dataFrame(sessionA, 5));
      buffer.feed(dataFrame(sessionA, 5));
      expect(buffer.uniqueSymbolCount, 100);
      expect(buffer.totalFramesSeen, 103);
      expect(buffer.droppedCount, 0);
    });

    test(
      'corrupt frame yields status error, drops, and leaves buffer intact',
      () {
        final buffer = FrameBuffer();
        buffer.feed(dataFrame(sessionA, 0));
        final corrupted = dataFrame(sessionA, 1);
        corrupted[40] ^= 0xff; // payload byte flip -> CRC32C mismatch
        expect(() => decodeFrame(corrupted), throwsA(isA<ProtocolError>()));
        final result = buffer.feed(corrupted);
        expect(result.status, FeedStatus.error);
        expect(result.frame, isNull);
        expect(result.isNewSession, isFalse);
        expect(buffer.droppedCount, 1);
        expect(buffer.totalFramesSeen, 1);
        expect(buffer.uniqueSymbolCount, 1);
        expect(buffer.symbols(), [payloadFor(0)]);
      },
    );

    test('new sessionId mid-stream resets and latches (isNewSession true)', () {
      final buffer = FrameBuffer();
      for (var esi = 0; esi < 30; esi++) {
        buffer.feed(dataFrame(sessionA, esi));
      }
      final result = buffer.feed(dataFrame(sessionB, 0));
      expect(result.status, FeedStatus.ok);
      expect(result.isNewSession, isTrue);
      expect(buffer.sessionId, sessionB);
      expect(buffer.uniqueSymbolCount, 1);
      expect(buffer.totalFramesSeen, 31);
      expect(buffer.symbols(), [payloadFor(0)]);
    });

    test(
      'META frame parses metadata; different-session META switches session',
      () {
        final buffer = FrameBuffer();
        final resultA = buffer.feed(metaFrame(sessionA));
        expect(resultA.status, FeedStatus.ok);
        expect(resultA.meta, isNotNull);
        expect(resultA.meta!.sessionId, sessionA);
        expect(buffer.metadata, isNotNull);
        expect(buffer.metadata!.sessionId, sessionA);
        expect(buffer.sessionId, sessionA);
        expect(buffer.k, 100);

        buffer.feed(dataFrame(sessionA, 7));
        expect(buffer.uniqueSymbolCount, 1);

        final resultB = buffer.feed(metaFrame(sessionB));
        expect(resultB.status, FeedStatus.ok);
        expect(resultB.isNewSession, isTrue);
        expect(resultB.meta!.sessionId, sessionB);
        expect(buffer.sessionId, sessionB);
        expect(buffer.metadata!.sessionId, sessionB);
        expect(buffer.uniqueSymbolCount, 0); // buffer reset on session switch
      },
    );

    test('META parse failure yields status dropped and drops', () {
      final buffer = FrameBuffer();
      final badJson = encodeFrame(
        Frame(
          type: typeMeta,
          sessionId: sessionA,
          esi: 0,
          k: 100,
          totalLen: 0,
          flags: 0,
          payload: Uint8List.fromList('not json'.codeUnits),
        ),
      );
      final result = buffer.feed(badJson);
      expect(result.status, FeedStatus.dropped);
      expect(result.meta, isNull);
      expect(buffer.droppedCount, 1);
      expect(buffer.totalFramesSeen, 1);
      expect(buffer.uniqueSymbolCount, 0);

      // Header sessionId differs from the metadata payload sessionId.
      final mismatched = encodeFrame(
        Frame(
          type: typeMeta,
          sessionId: sessionB,
          esi: 0,
          k: 100,
          totalLen: 0,
          flags: 0,
          payload: buildMetadataPayload(
            TransferMetadata(
              magic: metaMagic,
              protoVer: protoVersion,
              sessionId: sessionA,
              filename: 'fake.bin',
              mime: 'application/octet-stream',
              totalSize: 0,
              compressedSize: 0,
              compressed: false,
              k: 100,
              symbolSize: 1028,
              mtu: 1028,
              fileSHA256: 'a' * 64,
              flags: 0,
            ),
          ),
        ),
      );
      expect(buffer.feed(mismatched).status, FeedStatus.dropped);
      expect(buffer.droppedCount, 2);
      expect(buffer.metadata, isNull);
    });

    test('reset clears session state but keeps cumulative counters', () {
      final buffer = FrameBuffer();
      buffer.feed(dataFrame(sessionA, 0));
      buffer.feed(dataFrame(sessionA, 1));
      expect(buffer.totalFramesSeen, 2);
      buffer.reset();
      expect(buffer.uniqueSymbolCount, 0);
      expect(buffer.sessionId, isNull);
      expect(buffer.metadata, isNull);
      expect(buffer.k, isNull);
      expect(buffer.totalFramesSeen, 2);
      expect(buffer.droppedCount, 0);
    });

    test('eviction: source symbols never evicted, oldest repair goes first', () {
      final buffer = FrameBuffer(repairBudget: 10);
      for (var esi = 0; esi < 10; esi++) {
        buffer.feed(dataFrame(sessionA, esi, k: 10));
      }
      for (var esi = 10; esi < 35; esi++) {
        buffer.feed(dataFrame(sessionA, esi, k: 10));
      }
      // maxSymbols = k + budget = 20; 35 fed -> 15 oldest repair evicted.
      expect(buffer.uniqueSymbolCount, 20);
      final esis = buffer.symbolEsiSet();
      // Source symbols (esi < k) are never evicted.
      expect({for (var i = 0; i < 10; i++) i}.difference(esis), isEmpty);
      // Oldest repair (10..24) evicted first; newest repair (25..34) retained.
      expect({for (var i = 25; i < 35; i++) i}.difference(esis), isEmpty);
    });

    test('eviction: default budget keeps all symbols below the cap', () {
      final buffer = FrameBuffer();
      for (var esi = 0; esi < 10; esi++) {
        buffer.feed(dataFrame(sessionA, esi, k: 10));
      }
      for (var esi = 10; esi < 35; esi++) {
        buffer.feed(dataFrame(sessionA, esi, k: 10));
      }
      // maxSymbols = k + floor(k * 0.3) + 1000 = 1013; nothing evicted.
      expect(buffer.uniqueSymbolCount, 35);
      final esis = buffer.symbolEsiSet();
      expect({for (var i = 0; i < 10; i++) i}.difference(esis), isEmpty);
      expect({for (var i = 10; i < 35; i++) i}.difference(esis), isEmpty);
    });
  });
}

File: tests/unit/frames.test.ts
import { describe, expect, it } from 'vitest'
import { SplitMix32 } from '../../src/codec/fountain/rng'
import { META_MAGIC, TYPE_DATA } from '../../src/protocol/constants'
import { buildMetadataFrame, type Metadata } from '../../src/protocol/metadata'
import { encodeFrame, generateSessionId, type Frame } from '../../src/protocol/wire'
import { FrameBuffer, type FeedResult } from '../../src/receiver/frames'

const PAYLOAD_LEN = 64

/** Deterministic pseudo-random payload bytes (splitmix32, seeded per esi). */
function payloadFor(esi: number): Uint8Array {
  const rng = new SplitMix32(1000 + esi)
  const bytes = new Uint8Array(PAYLOAD_LEN)
  for (let i = 0; i < PAYLOAD_LEN; i++) {
    bytes[i] = rng.int(256)
  }
  return bytes
}

function dataFrame(sessionId: string, esi: number, k: number): Frame {
  return {
    type: TYPE_DATA,
    sessionId,
    esi,
    k,
    totalLen: k * PAYLOAD_LEN,
    flags: 0,
    payload: payloadFor(esi),
  }
}

function metaFor(sessionId: string, k: number, filename = 'test.bin'): Metadata {
  return {
    magic: META_MAGIC,
    protoVer: 1,
    sessionId,
    filename,
    mime: 'application/octet-stream',
    totalSize: k * PAYLOAD_LEN,
    compressedSize: 0,
    compressed: false,
    k,
    symbolSize: PAYLOAD_LEN,
    mtu: 100,
    fileSHA256: 'ab'.repeat(32),
    flags: 0,
  }
}

/** Deterministic non-identity permutation of esi 0..count-1 (evens, then odds). */
function shuffledEsi(count: number): number[] {
  const order: number[] = []
  for (let i = 0; i < count; i += 2) order.push(i)
  for (let i = 1; i < count; i += 2) order.push(i)
  return order
}

function feedData(buffer: FrameBuffer, sessionId: string, k: number, esi: number): FeedResult {
  return buffer.feed(encodeFrame(dataFrame(sessionId, esi, k)))
}

describe('FrameBuffer', () => {
  it('collects 100 in-order DATA frames with no drops', () => {
    const sessionId = generateSessionId()
    const k = 100
    const buffer = new FrameBuffer()

    for (let esi = 0; esi < k; esi++) {
      expect(feedData(buffer, sessionId, k, esi).status).toBe('ok')
    }

    expect(buffer.uniqueSymbolCount).toBe(100)
    expect(buffer.droppedCount).toBe(0)
    expect(buffer.totalFramesSeen).toBe(100)
    expect(buffer.sessionId).toBe(sessionId)
    expect(buffer.k).toBe(100)
    expect(buffer.symbols()).toHaveLength(100)
    expect(buffer.symbols()[0]).toEqual(payloadFor(0))
    expect(buffer.symbols()[99]).toEqual(payloadFor(99))
    const esiSet = buffer.symbolEsiSet()
    expect(esiSet.has(0)).toBe(true)
    expect(esiSet.has(99)).toBe(true)
  })

  it('collects the same 100 symbols when fed out of order', () => {
    const sessionId = generateSessionId()
    const k = 100
    const buffer = new FrameBuffer()

    for (const esi of shuffledEsi(k)) {
      feedData(buffer, sessionId, k, esi)
    }

    expect(buffer.uniqueSymbolCount).toBe(100)
    expect(buffer.totalFramesSeen).toBe(100)
    expect(buffer.symbols()).toHaveLength(100)
    expect(buffer.symbols()[0]).toEqual(payloadFor(0))
    expect(buffer.symbols()[99]).toEqual(payloadFor(99))
  })

  it('dedups repeated esi without losing frame accounting', () => {
    const sessionId = generateSessionId()
    const k = 100
    const buffer = new FrameBuffer()

    for (let esi = 0; esi < k; esi++) feedData(buffer, sessionId, k, esi)
    for (let i = 0; i < 3; i++) feedData(buffer, sessionId, k, 5)

    expect(buffer.uniqueSymbolCount).toBe(100)
    expect(buffer.totalFramesSeen).toBe(103)
    expect(buffer.droppedCount).toBe(0)
  })

  it('treats a corrupted frame as an error without throwing or mutating state', () => {
    const sessionId = generateSessionId()
    const k = 100
    const buffer = new FrameBuffer()
    for (let esi = 0; esi < 10; esi++) feedData(buffer, sessionId, k, esi)

    const corrupted = encodeFrame(dataFrame(sessionId, 50, k))
    corrupted[30 + 5] = 0xff // flip a payload byte -> BAD_CRC

    const result = buffer.feed(corrupted)
    expect(result.status).toBe('error')
    expect(result.frame).toBeUndefined()
    expect(buffer.droppedCount).toBe(1)
    expect(buffer.uniqueSymbolCount).toBe(10)
    expect(buffer.sessionId).toBe(sessionId)

    // A valid frame after the corrupted one still lands.
    expect(feedData(buffer, sessionId, k, 50).status).toBe('ok')
    expect(buffer.uniqueSymbolCount).toBe(11)
  })

  it('starts fresh on a new session mid-stream', () => {
    const sessionA = generateSessionId()
    const sessionB = generateSessionId()
    const k = 100
    const buffer = new FrameBuffer()

    expect(buffer.feed(buildMetadataFrame(metaFor(sessionA, k))).status).toBe('ok')
    for (let esi = 0; esi < 30; esi++) feedData(buffer, sessionA, k, esi)
    expect(buffer.uniqueSymbolCount).toBe(30)

    const switchResult = feedData(buffer, sessionB, k, 7)
    expect(switchResult.status).toBe('ok')
    expect(switchResult.isNewSession).toBe(true)
    expect(buffer.sessionId).toBe(sessionB)
    expect(buffer.metadata).toBeUndefined()
    expect(buffer.uniqueSymbolCount).toBe(1)
  })

  it('parses META frames and latches the session', () => {
    const sessionId = generateSessionId()
    const k = 100
    const buffer = new FrameBuffer()
    const meta = metaFor(sessionId, k)

    const result = buffer.feed(buildMetadataFrame(meta))

    expect(result.status).toBe('ok')
    expect(result.isNewSession).toBe(true)
    expect(result.meta).toEqual(meta)
    expect(buffer.metadata).toEqual(meta)
    expect(buffer.sessionId).toBe(sessionId)
    expect(buffer.k).toBe(k)
  })

  it('treats a META frame with a new sessionId as a session switch', () => {
    const sessionA = generateSessionId()
    const sessionB = generateSessionId()
    const buffer = new FrameBuffer()
    for (let esi = 0; esi < 10; esi++) feedData(buffer, sessionA, 100, esi)
    const metaB = metaFor(sessionB, 50)

    const result = buffer.feed(buildMetadataFrame(metaB))

    expect(result.isNewSession).toBe(true)
    expect(buffer.sessionId).toBe(sessionB)
    expect(buffer.metadata).toEqual(metaB)
    expect(buffer.uniqueSymbolCount).toBe(0)
    expect(buffer.k).toBe(50)
  })

  it('starts empty', () => {
    const buffer = new FrameBuffer()
    expect(buffer.metadata).toBeUndefined()
    expect(buffer.sessionId).toBeUndefined()
    expect(buffer.k).toBeUndefined()
    expect(buffer.uniqueSymbolCount).toBe(0)
    expect(buffer.totalFramesSeen).toBe(0)
    expect(buffer.droppedCount).toBe(0)
    expect(buffer.symbols()).toEqual([])
    expect(buffer.symbolEsiSet().size).toBe(0)
  })

  it('exposes k from DATA frame headers before any META arrives', () => {
    const buffer = new FrameBuffer()
    const sessionId = generateSessionId()
    feedData(buffer, sessionId, 100, 3)
    expect(buffer.k).toBe(100)
  })

  it('keeps the total bounded with a default repair budget', () => {
    const sessionId = generateSessionId()
    const k = 10
    const buffer = new FrameBuffer()

    for (let esi = 0; esi < 10; esi++) feedData(buffer, sessionId, k, esi) // source
    for (let esi = 10; esi < 35; esi++) feedData(buffer, sessionId, k, esi) // repair

    for (let esi = 0; esi < 10; esi++) {
      expect(buffer.symbolEsiSet().has(esi)).toBe(true)
    }
    expect(buffer.uniqueSymbolCount).toBe(35)
    expect(buffer.uniqueSymbolCount).toBeLessThanOrEqual(k * 1.3 + 1000)
  })

  it('evicts the oldest repair symbols when the repair budget is exceeded', () => {
    const sessionId = generateSessionId()
    const k = 10
    const buffer = new FrameBuffer({ repairBudget: 5 }) // cap at k + 5 = 15 symbols

    for (let esi = 0; esi < 10; esi++) feedData(buffer, sessionId, k, esi)
    for (let esi = 10; esi < 25; esi++) feedData(buffer, sessionId, k, esi)

    const esiSet = buffer.symbolEsiSet()
    for (let esi = 0; esi < 10; esi++) expect(esiSet.has(esi)).toBe(true) // source retained
    for (let esi = 10; esi < 20; esi++) expect(esiSet.has(esi)).toBe(false) // oldest repair evicted
    for (let esi = 20; esi < 25; esi++) expect(esiSet.has(esi)).toBe(true) // newest repair kept
    expect(buffer.uniqueSymbolCount).toBe(15)
  })

  it('reset() clears session state while keeping cumulative counters', () => {
    const sessionId = generateSessionId()
    const buffer = new FrameBuffer()
    feedData(buffer, sessionId, 100, 0)
    feedData(buffer, sessionId, 100, 0)
    buffer.feed(new Uint8Array(8)) // too short to be a frame -> dropped

    buffer.reset()

    expect(buffer.sessionId).toBeUndefined()
    expect(buffer.metadata).toBeUndefined()
    expect(buffer.k).toBeUndefined()
    expect(buffer.uniqueSymbolCount).toBe(0)
    expect(buffer.symbols()).toEqual([])
    expect(buffer.totalFramesSeen).toBe(2)
    expect(buffer.droppedCount).toBe(1)
  })
})
